File: stats.py
class StatisticsCalculator:
    def __init__(self, numbers):
        self.numbers = numbers

    def mean(self):
        return sum(self.numbers) / len(self.numbers)

    def median(self):
        nums = sorted(self.numbers)
        n = len(nums)
        mid = n // 2
        if n % 2 == 0:
            return (nums[mid - 1] + nums[mid]) / 2
        return nums[mid]

    def mode(self):
        frequency = {}
        for num in self.numbers:
            frequency[num] = frequency.get(num, 0) + 1

        max_freq = max(frequency.values())
        return [num for num, freq in frequency.items() if freq == max_freq]


if __name__ == "__main__":
    data = [4, 1, 2, 2, 5, 3, 3]
    calc = StatisticsCalculator(data)

    print("Mean:", calc.mean())
    print("Median:", calc.median())
    print("Mode(s):", calc.mode())

File: stats.ml
let mean lst =
  let sum = List.fold_left ( + ) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 1 then
    float_of_int (List.nth sorted (n / 2))
  else
    let a = List.nth sorted (n/2 - 1) in
    let b = List.nth sorted (n/2) in
    (float_of_int a +. float_of_int b) /. 2.0

let mode lst =
  let sorted = List.sort compare lst in
  let rec count current count_max current_count modes = function
    | [] ->
        if current_count = count_max then current :: modes
        else if current_count > count_max then [current]
        else modes
    | x :: xs ->
        if x = current then
          count current count_max (current_count + 1) modes xs
        else
          let new_modes =
            if current_count = count_max then current :: modes
            else if current_count > count_max then [current]
            else modes
          in
          count x (max count_max current_count) 1 new_modes xs
  in
  match sorted with
  | [] -> []
  | x :: xs -> count x 1 1 [] xs

let () =
  let data = [4; 1; 2; 2; 5; 3; 3] in
  Printf.printf "Mean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);

  let m = mode data in
  Printf.printf "Mode(s): ";
  List.iter (fun x -> Printf.printf "%d " x) m;
  print_endline ""
